src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,233 @@
+import { Dialog, Transition } from '@headlessui/react';
+import clsx from 'clsx';
+import React, { MutableRefObject, ReactNode } from 'react';
+import styles from './Modal.module.css';
+import { Icon } from '../Icon/Icon';
+import { ModalBody } from '../ModalBody/ModalBody';
+import {
+  ModalFooter,
+  Props as ModalFooterProps,
+} from '../ModalFooter/ModalFooter';
+import {
+  ModalHeader,
+  Props as ModalHeaderProps,
+} from '../ModalHeader/ModalHeader';
+import { ModalStepper } from '../ModalStepper/ModalStepper';
+import { ModalTitle } from '../ModalTitle/ModalTitle';
+
+type Variant = 'brand' | undefined;
+
+type ModalContentProps = {
+  /**
+   * Optional aria-label for the modal.
+   *
+   * If undefined, the headingText of the Modal.Header will be used.
+   * If there is no Modal.Header, an ariaLabel is required.
+   */
+  ariaLabel?: string;
+  /**
+   * Additional classnames passed in for styling.
+   */
+  className?: string;
+  /**
+   * Contents of the modal.
+   */
+  children: ReactNode;
+  /**
+   * Hides the close button in the top right of the modal.
+   */
+  hideCloseButton?: boolean;
+  /**
+   * A ref to an element that should receive focus when the modal first opens.
+   *
+   * If undefined, the first focusable element (usually the close button) will be used.
+   *
+   * @example
+   * const inputFieldRef = useRef();
+   *
+   * <Modal initialFocus={inputFieldRef}>
+   *   ...
+   *   <InputField ref={inputFieldRef} />
+   * </Modal>
+   */
+  initialFocus?: MutableRefObject<HTMLElement | null>;
+  /**
+   * Toggles scrollable variant of the modal. If modal is scrollable, footer is not, and vice versa.
+   * Defaults to false since modal default is not scrollable.
+   * Also adds border and shadow to the footer indicate sticky status.
+   */
+  isScrollable?: boolean;
+  /**
+   * Method called when the close button is clicked. Use this to hide the modal.
+   *
+   * This is required even if you don't have a close button so the ESC key can close the modal.
+   */
+  onClose: () => void;
+  /**
+   * Color variants of the modal.
+   */
+  variant?: Variant;
+};
+
+type ModalProps = ModalContentProps & {
+  /**
+   * Whether or not the modal is visible.
+   */
+  open: boolean;
+};
+
+type Context = {
+  isScrollable?: boolean;
+  variant?: Variant;
+};
+
+const ModalContext = React.createContext<Context>({});
+
+function childrenHaveModalTitle(children?: ReactNode): boolean {
+  const childrenArray = React.Children.toArray(children);
+  return childrenArray.some((child) => {
+    if (typeof child === 'string' || typeof child === 'number') {
+      return false;
+    } else if (
+      'props' in child &&
+      child.type &&
+      typeof child.type !== 'string' &&
+      child.type?.name === 'ModalTitle'
+    ) {
+      return true;
+    } else if ('props' in child && child.props.children) {
+      return childrenHaveModalTitle(child.props.children);
+    }
+    return false;
+  });
+}
```

Comment:
I'm not sure how common it is, but I've seen at least one instance of a Heading + Subheading in the modal title. If I'm reading this right, this allows for that so long as at least one of the children is a Modal.Title, right? Purely for example, something like
```tsx
<div>
  <Modal.Title>Heading 1</Modal.Title>
  <Heading as="h2">Heading 2</Heading>
</div>